src/override/memcpy.cc

```diff
@@ -0,0 +1,230 @@
+#include "override.h"
+
+#include <errno.h>
+#include <stdio.h>
+#include <string.h>
+#if __has_include(<xlocale.h>)
+#  include <xlocale.h>
+#endif
+
+using namespace snmalloc;
+
+// glibc lacks snprintf_l
+#ifdef __linux__
+#  define snprintf_l(buf, size, loc, msg, ...) \
+    snprintf(buf, size, msg, __VA_ARGS__)
+// Windows has it with an underscore prefix
+#elif defined(_MSC_VER)
+#  define snprintf_l(buf, size, loc, msg, ...) \
+    _snprintf_l(buf, size, msg, loc, __VA_ARGS__)
+#endif
+
+namespace
+{
+  /**
+   * Should we check loads?  This defaults to on in debug builds, off in
+   * release (store-only checks)
+   */
+  static constexpr bool CheckReads =
+#ifdef SNMALLOC_CHECK_LOADS
+    SNMALLOC_CHECK_LOADS
+#else
+#  ifdef NDEBUG
+    false
+#  else
+    true
+#  endif
+#endif
+    ;
+
+  /**
+   * Should we fail fast when we encounter an error?  With this set to true, we
+   * just issue a trap instruction and crash the process once we detect an
+   * error. With it set to false we print a helpful error message and then crash
+   * the process.  The process may be in an undefined state by the time the
+   * check fails, so there are potentially security implications to turning this
+   * off. It defaults to true for debug builds, false for release builds.
+   */
+  static constexpr bool FailFast =
+#ifdef SNMALLOC_FAIL_FAST
+    SNMALLOC_FAIL_FAST
+#else
+#  ifdef NDEBUG
+    true
+#  else
+    false
+#  endif
+#endif
+    ;
+
+  /**
+   * The largest register size that we can use for loads and stores.  These
+   * types are expected to work for overlapping copies: we can always load them
+   * into a register and store them.  Note that this is at the C abstract
+   * machine level: the compiler may spill temporaries to the stack, just not
+   * to the source or destination object.
+   */
+  static constexpr size_t LargestRegisterSize =
+#ifdef __AVX__
+    32
+#elif defined(__SSE__)
+    16
+#else
+    sizeof(uint64_t)
+#endif
+    ;
+
+  /**
+   * Copy a single element of a specified size.  Uses a compiler builtin that
+   * expands to a single load and store.
+   */
+  template<size_t Size>
+  SNMALLOC_FAST_PATH inline void copy_one(void* dst, const void* src)
+  {
+#if __has_builtin(__builtin_memcpy_inline)
+    __builtin_memcpy_inline(dst, src, Size);
+#else
+    // Define a structure of size `Size` that has alignment 1 and a default
+    // copy-assignment operator.  We can then copy the data as this type.  The
+    // compiler knows the exact width and so will generate the correct wide
+    // instruction for us (clang 10 and gcc 12 both generate movups for the
+    // 16-byte version of this when targeting SSE.
+    struct Block
+    {
+      char data[Size];
+    };
+    auto* d = static_cast<Block*>(dst);
+    auto* s = static_cast<const Block*>(src);
+    *d = *s;
+#endif
+  }
+
+  SNMALLOC_SLOW_PATH SNMALLOC_UNUSED_FUNCTION void crashWithMessage
```

Comment:
Could be if we have other consumers of it.  If we want to add more things with bounds checks, it definitely should be.

Reply:
Well, there are a bunch of things that do checks

* freeing the start of object
* corruption of free list

So, giving those nicer error message would be good.  But clearly a different PR.  Let's leave that refactoring to then.

Reply:
Sounds good.  This function is quite specialised because it's trying to keep everything related to preparing the message in the slow-path function, so that we can just bail out and do all of the complicated formatting things in a separate function.  It would probably need to be refactored to be a generic that took a lambda that did the formatting if we wanted something more general, ideally with the fast-path and slow-path bits both being generic.  I'll have a think about what that should look like.